pypicloud/storage/azure_blob.py

```diff
@@ -0,0 +1,180 @@
+            "storage.azure_storage_account_name"
+        )
+        if kwargs["azure_storage_account_name"] is None:
+            raise ValueError(
+                "You must specify the 'storage.azure_storage_account_name'"
+            )
+
+        kwargs["azure_storage_account_key"] = settings.get(
+            "storage.azure_storage_account_key"
+        )
+        if kwargs["azure_storage_account_key"] is None:
+            raise ValueError("You must specify the 'storage.azure_storage_account_key'")
+
+        kwargs["azure_storage_container_name"] = settings.get(
+            "storage.azure_storage_container_name"
+        )
+        if kwargs["azure_storage_container_name"] is None:
+            raise ValueError(
+                "You must specify the 'storage.azure_storage_container_name'"
+            )
+
+        return kwargs
+
+    def _generate_url(self, package):
+        path = self.get_path(package)
+
+        url_params = generate_blob_sas(
+            account_name=self.azure_storage_account_name,
+            container_name=self.azure_storage_container_name,
+            blob_name=path,
+            account_key=self.azure_storage_account_key,
+            permission=BlobSasPermissions(read=True),
+            expiry=datetime.now() + timedelta(seconds=self.expire_after),
+            protocol="https",
+        )
+
+        url = "{}/{}/{}?{}".format(
+            self.azure_storage_account_url,
+            self.azure_storage_container_name,
+            path,
+            url_params,
+        )
+        return url
+
+    def download_response(self, package):
+        return HTTPFound(location=self._generate_url(package))
+
+    def list(self, factory=Package):
+        # List does not return metadata :(
+        for blob_properties in self.container_client.list_blobs(
+            name_starts_with=self.path_prefix
+        ):
+            blob_client = self.container_client.get_blob_client(
+                blob=blob_properties.name
+            )
+            metadata = blob_client.get_blob_properties()
+
+            yield factory(
+                metadata.metadata["name"],
+                metadata.metadata["version"],
+                posixpath.basename(blob_properties.name),
+                blob_properties.last_modified,
+                path=blob_properties.name,
+                **Package.read_metadata(metadata.metadata)
+            )
+
+    def get_path(self, package):
+        """ Get the fully-qualified bucket path for a package """
+        if "path" not in package.data:
+            package.data["path"] = (
+                self.path_prefix + package.name + "/" + package.filename
+            )
+        return package.data["path"]
+
+    def upload(self, package, datastream):
+        path = self.get_path(package)
+
+        metadata = package.get_metadata()
+        metadata["name"] = package.name
+        metadata["version"] = package.version
+        self._normalize_metadata(metadata)
```

Comment:
Do you know if this step is actually required for azure-blob storage? This was added in #197 and seems to be because of a limitation of the AWS REST API (it doesn't support non-ascii text). If this *is* also required by azure, could you move the method into the util.py file so both storage classes can use the same function?

Reply:
Yep it gets very unhappy when you pass it utf8 characters, will move